src/compiler/checker.ts

```diff
@@ -6404,6 +6404,22 @@ namespace ts {
             let container = getThisContainer(node, /* includeArrowFunctions */ true);
             let needToCaptureLexicalThis = false;
 
+
+            if (container.kind === SyntaxKind.Constructor) {
+                // Keep track of whether we have seen "super" before encounter "this" so that
+                // we can report appropriate error later in checkConstructorDeclaration
+                // We have to do the check here to make sure we won't give false error when
+                // "this" is used in arrow functions
+                // For example:
+                //      constructor() {
+                //          (()=>this);  // No Error
+                //          super();
+                //      }
+                if ((<ConstructorDeclaration>container).hasSeenSuperBeforeThis === undefined) {

```

Comment:
Would it be better to make this a value on `NodeCheckFlags`, rather than a property on `ConstructorDeclaration`?